causallearn/utils/cit.py

```diff
@@ -15,78 +15,142 @@
+    method: str, in ["fisherz", "mv_fisherz", "mc_fisherz", "kci", "chisq", "gsq"]
+    kwargs: placeholder for future arguments, or for KCI specific arguments now
+    '''
+    if method == fisherz:
+        return FisherZ(data, **kwargs)
+    elif method == kci:
+        return KCI(data, **kwargs)
+    elif method in [chisq, gsq]:
+        return Chisq_or_Gsq(data, method_name=method, **kwargs)
+    elif method == mv_fisherz:
+        return MV_FisherZ(data, **kwargs)
+    elif method == mc_fisherz:
+        return MC_FisherZ(data, **kwargs)
+    else:
+        raise ValueError("Unknown method: {}".format(method))
+
+class CIT_Base(object):
+    # Base class for CIT, contains basic operations for input check and caching, etc.
+    def __init__(self, data, cache_path=None, **kwargs):
         '''
         Parameters
         ----------
-        data: numpy.ndarray of shape (n_samples, n_features)
-        method: str, in ["fisherz", "mv_fisherz", "mc_fisherz", "kci", "chisq", "gsq"]
-        kwargs: placeholder for future arguments, or for KCI specific arguments now
+        data: data matrix, np.ndarray, in shape (n_samples, n_features)
+        cache_path: str, path to save cache .json file. default as None (no io to local file).
+        kwargs: for future extension.
         '''
+        assert isinstance(data, np.ndarray), "Input data must be a numpy array."
         self.data = data
-        self.data_hash = hash(str(data))
+        self.data_hash = hashlib.md5(str(data).encode('utf-8')).hexdigest()
         self.sample_size, self.num_features = data.shape
-        self.method = method
-        self.pvalue_cache = {}
-
-        if method == 'kci':
-            # parse kwargs contained in the KCI method
-            kci_ui_kwargs = {k: v for k, v in kwargs.items() if k in
-                             ['kernelX', 'kernelY', 'null_ss', 'approx', 'est_width', 'polyd', 'kwidthx', 'kwidthy']}
-            kci_ci_kwargs = {k: v for k, v in kwargs.items() if k in
-                             ['kernelX', 'kernelY', 'kernelZ', 'null_ss', 'approx', 'use_gp', 'est_width', 'polyd',
-                              'kwidthx', 'kwidthy', 'kwidthz']}
-            self.kci_ui = KCI_UInd(**kci_ui_kwargs)
-            self.kci_ci = KCI_CInd(**kci_ci_kwargs)
-        elif method in ['fisherz', 'mv_fisherz', 'mc_fisherz']:
-            self.correlation_matrix = np.corrcoef(data.T)
-        elif method in ['chisq', 'gsq']:
-            def _unique(column):
-                return np.unique(column, return_inverse=True)[1]
-            self.data = np.apply_along_axis(_unique, 0, self.data).astype(np.int64)
-            self.data_hash = hash(str(self.data))
-            self.cardinalities = np.max(self.data, axis=0) + 1
+        self.cache_path = cache_path
+        self.SAVE_CACHE_CYCLE_SECONDS = 30
+        self.last_time_cache_saved = time.time()
+        self.pvalue_cache = {'data_hash': self.data_hash}
+        if not cache_path is None:
```

Comment:
```suggestion
        if cache_path is not None:
```

check https://stackoverflow.com/questions/2710940/python-if-x-is-not-none-or-if-not-x-is-none

Reply:
Wow, thanks for this!!! Marked!

I usually get lost when writing something like this...